sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -315,7 +315,7 @@ private static Expression castOutput(Expression value, FieldType toType) {
   private static Expression castOutputTime(Expression value, FieldType toType) {
     Expression valueDateTime = value;
 
-    // First, convert to millis (except for DATE type)
+    // First, convert to millis (except for DATE/TIME type)
```

Comment:
All the current failing tests are due to the code below (`TIME` case). Note the comment says "convert to millis", which means the values are in milliseconds. Previously it will hit this code path: `Expressions.new_(Instant.class, valueDateTime);`, which corresponds to calling `new Instant(long)`, which takes milliseconds as a long as the input parameter. Now you have updated the code so it will hit calling `LocalTime.ofNanoOfDay()` which takes nanosecond. The fix will be adding 1 LOC: `valueDateTime = Expressions.multiply(valueDateTime, Expressions.constant(1000000L))`.

Reply:
Another problem is that `LocalTime.ofNanoOfDay()` takes a `long` instead of `Long`. So when the `valueDateTime` is of type `Long` the expression cannot be resolved. The fix is to add:
```
if (value.getType() == Long.class) {
  valueDateTime = Expressions.unbox(valueDateTime);
}
```
(The same fix should probably be added to `DATE` case as well. It has not failed simply because no test has hit that case.)

Reply:
Got it. Thanks for figuring it out.